From: Michael Schutte <michi@debian.org>
Subject: openvt: Improved error handling and don’t close consfd twice
Forwarded: no
Bug-Debian: http://bugs.debian.org/190387
Bug-Debian: http://bugs.debian.org/630108

Two unrelated fixes:

 * Propagate child exit codes in “openvt -w”.

 * consfd stores the file descriptor referring to the console.  In some
   cases, this will be one of the stdio fds, which are closed in the
   child process.  Avoid closing it again.

---
 src/openvt.c |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

Index: devel/src/openvt.c
===================================================================
--- devel.orig/src/openvt.c	2011-07-22 21:45:00.000000000 +0200
+++ devel/src/openvt.c	2011-07-22 21:45:04.000000000 +0200
@@ -385,7 +385,8 @@
 		close(0);
 		close(1);
 		close(2);
-		close(consfd);
+		if (consfd > 2)
+			close(consfd);
 
 		if ((dup2(fd, 0) == -1) || (dup2(fd, 1) == -1) || (dup2(fd, 2) == -1))
 			openvt_fatal(1, errno, "dup");
@@ -408,7 +409,6 @@
 	if (do_wait) {
 		int retval = 0; /* actual value returned form process */
 
-		wait(NULL);
 		waitpid(pid, &retval, 0);
 
 		if (show) {	/* Switch back... */
@@ -425,7 +425,10 @@
 
 		/* if all our stuff went right, we want to return the exit code of the command we ran
 		   super vital for scripting loops etc */
-		return(retval);
+		if (WIFEXITED(retval))
+			return WEXITSTATUS(retval);
+		else
+			return 127;
 	}
 
 	return EXIT_SUCCESS;
